osbuild/util/fscache.py

```diff
@@ -0,0 +1,1008 @@
+        time. Hence, a cache miss only reflects the state of the cache at a
+        particular time under a particular lock.
+        """
+
+    # static parameters
+    _dirname_data = "data"
+    _dirname_objects = "objects"
+    _dirname_stage = "stage"
+    _filename_cache_info = "cache.info"
+    _filename_cache_lock = "cache.lock"
+    _filename_cache_size = "cache.size"
+    _filename_object_info = "object.info"
+    _filename_object_lock = "object.lock"
+
+    # constant properties
+    _appid: str
+    _libc: linux.Libc
+    _path_cache: str
+
+    # context-manager properties
+    _active: bool
+    _bootid: Optional[str]
+    _lock: Optional[int]
+    _info: FsCacheInfo
+    _info_maximum_size: int
+
+    def __init__(self, appid: str, path_cache: str):
+        """Create File System Cache
+
+        This creates a new file-system cache. It does not create the cache, nor
+        access any of its content. You must enter its context-manager to prepare
+        the cache for access. Any access outside of a context-manager will raise
+        an assertion error, unless explicitly stated otherwise.
+
+        Parameters:
+        -----------
+        appid
+            The application-ID of the caller. This can be any random string. It
+            is used to initialize the application-specific boot-ID used to tag
+            caches and detect whether an entry was created during the same boot.
+        path_cache
+            The path to the cache directory. The directory (and the path to it)
+            is created if it does not exist.
+        """
+
+        self._appid = appid
+        self._libc = linux.Libc.default()
+        self._path_cache = path_cache
+
+        self._active = False
+        self._bootid = None
+        self._lock = None
+        self._info = FsCacheInfo()
+        self._info_maximum_size = 0
+
+    @staticmethod
+    def _calculate_size(path_target: str) -> int:
+        """Calculate total size of a directory tree
+
+        Calculate the total amount of storage required for a directory tree in
+        bytes. This does not account for metadata, but only for stored file
+        content.
+
+        Parameters:
+        -----------
+        path_target
+            File-system path to the directory to operate on.
+        """
+
+        return sum(
+            os.lstat(
+                os.path.join(path, f)
+            ).st_size for path, dirs, files in os.walk(
+                path_target
+            ) for f in files
+        )
+
+    def __fspath__(self) -> str:
+        """Return cache path
+
+        Return the path to this cache as provided to the constructor of the
+        cache. No conversions are applied, so the path is absolute if the
+        path as provided by the caller was absolute, and vice-versa.
+
+        This is part of the `os.PathLike` interface. See its documentation.
+        """
+
+        return self._path_cache
```

Comment:
```suggestion
        return os.fspath(self._path_cache)
```

The description for this special method is `The method should only return a str or bytes object, with the preference being for str.`. But if someone passed a `os.PathLike` to the `FsCache` you return that here as-is; everything else will work but then downstream consumers `FsCache` won't (`E       TypeError: expected FsCache.__fspath__() to return str or bytes, not PosixPath`)

Reply:
`_path_cache` is a `str`, and `__init__()` also requires a `str`. Would you prefer if we take `PathLike` in `__init__()`? I would still do the conversion early and store it as `str` in the cache.

Reply:
I am routing through what `ObjectStore` gets and in certain tests this is a `PathLib.Path`. I had a type (`osbuild.types.PathLike`) that was `Union[bytes, string, os.PathLike]` but it got destroyed by the mypy PR and I didn't notice and now it is basically useless.
Doing the conversion early sounds totally fine.